testkit/stepped.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testkit
+
+import (
+	"testing"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/breakpoint"
+	"github.com/stretchr/testify/require"
+)
+
+var errCommandRunFailed = errors.New("command run failed")
+
+type steppedTestkitMsgType int
+
+const (
+	msgTpCmdContinue steppedTestkitMsgType = iota
+	msgTpCmdStopOnBreakPoint
+	msgTpCmdDone
+)
+
+type steppedTestKitMsg struct {
+	tp  steppedTestkitMsgType
+	val any
+}
+
+type steppedTestKitMsgChan chan *steppedTestKitMsg
+
+func (ch steppedTestKitMsgChan) sendMsg(tp steppedTestkitMsgType, val any) error {
+	select {
+	case ch <- &steppedTestKitMsg{tp: tp, val: val}:
+		return nil
+	case <-time.After(time.Second * 10):
+		return errors.New("send msg timeout")
+	}
+}
+
+func (ch steppedTestKitMsgChan) sendMsgWithType(tp steppedTestkitMsgType) error {
+	return ch.sendMsg(tp, nil)
+}
+
+func (ch steppedTestKitMsgChan) recvMsg() (*steppedTestKitMsg, error) {
+	select {
+	case msg := <-ch:
+		return msg, nil
+	case <-time.After(time.Second * 10):
```

Comment:
Extract constant timeout?

Reply:
Yes, it is for easy... Because currently there is no requirements to set timeout manually and most sql commands in unit tests executes very fast.

Reply:
Just for two identical timeout value(another one at L49), it is not necessarily to be exposed.

Reply:
Do you mean make it as a constant?

Reply:
I've add a constant now `var defaultChanTimeout = time.Second * 10`